### PR TITLE
Supporting reduction into scalar, for float and half

### LINKED ISSUE #491: ScheduleReduction cannot reduce into a single number
## 🐛 Bug

The current `scheduleReduction` implementation always assumes a non-reduction axis present in the reduction output tensor. This is not the case when reducing to a single scalar value.

## To Reproduce
```python
x = torch.randn([4, 5, 6], dtype=dtype, device=device)
def t(x: torch.Tensor):
  o = torch.add(x, 0)
  o = torch.sum(o, dim=[0,1,2])
  return o
t_jit = torch.jit.script(t)
jit_o = t_jit(x)
jit_o = t_jit(x)
o = t(x)
self.assertEqual(o.dtype, jit_o.dtype)
self.assertEqual(o, jit_o)
self.assertGraphContains(t_jit.graph_for(x), FUSION_GUARD)
```

### PR DESCRIPTION
**Pull request by shmsong**
Fixes #491 

- Minor changes on the `fastest_dim` part of scheduleReduction to accomodate reductions with 0 output dims.

- Updated `lower_unroll` pass to predicate on top level expressions that write to global memory. This case happens with a half-typed reduction that ends up with a simple half-typed scalar. The final conversion in the kernel will not be inside any loop.

Example kernel without fix:

```cpp
__global__ void CUDAGeneratedKernel(Tensor<__half, 1> T0, Tensor<__half, 0> T3) {
    alignas(4) extern __shared__ char array[];
    void* shared_mem = array;
    float T2[1];
    T2[0] = float(0);
    float T4[1];
    T4[0] = float(0);
    for(size_t ki17 = 0; ki17 < (ceilDiv((ceilDiv(T0.size[0], 4)), blockDim.x)); ++ki17) {
      float T1[4];
      if ((((((ki17 * blockDim.x) + threadIdx.x) * 4) + (4 - 1)) < T0.size[0])) {
        for(size_t ki20 = 0; ki20 < 4; ++ki20) {
          T1[ki20]
             = __half2float(T0[(((((ki17 * blockDim.x) + threadIdx.x) * 4) + ki20) * T0.stride[0])]);
          T4[0]
            = T4[0]
            + T1[ki20];
        }
      } else {
        for(size_t ki20 = 0; ki20 < 4; ++ki20) {
          if ((((((ki17 * blockDim.x) + threadIdx.x) * 4) + ki20) < T0.size[0])) {
            T1[ki20]
               = __half2float(T0[(((((ki17 * blockDim.x) + threadIdx.x) * 4) + ki20) * T0.stride[0])]);
          }
          if ((((((ki17 * blockDim.x) + threadIdx.x) * 4) + ki20) < T0.size[0])) {
            T4[0]
              = T4[0]
              + T1[ki20];
          }
        }
      }
    }
    blockReduce<true, false, false>(
      T2[0],
      T4[0],
      [](float &a, float b) { a = a + b; },
      threadIdx,
      blockDim,
      static_cast<float*>(shared_mem),
      1,
      float(0));
    T3[0]
       = __float2half(T2[0]);
  }
```

The very last line `T3[0] = __float2half(T2[0]);` is the one that misses a thread predicate because it sits on the top level. 

Rparams covered by new test case `FusionReductionSchedulerNoODimShmoo`
```

===== Reduction Parameters ========
Inputs:
        Red Elems: 1 Red Outputs: 1 Red On Fastest Dim? 1
Reduction Characteristics:
        Multiple Reds Per Block? 1 Cross Block? 0 Cross Grid? 0
Recommended Blocking:
        GridX: 1 GridY: -1 BlckX: 1 BlckY: 1
====================================

===== Reduction Parameters ========
Inputs:
        Red Elems: 65536 Red Outputs: 1 Red On Fastest Dim? 1
Reduction Characteristics:
        Multiple Reds Per Block? 0 Cross Block? 1 Cross Grid? 0
Recommended Blocking:
        GridX: 1 GridY: -1 BlckX: 512 BlckY: 1
====================================

===== Reduction Parameters ========
Inputs:
        Red Elems: 262144 Red Outputs: 1 Red On Fastest Dim? 1
Reduction Characteristics:
        Multiple Reds Per Block? 0 Cross Block? 1 Cross Grid? 1
Recommended Blocking:
        GridX: 1 GridY: 32 BlckX: 512 BlckY: 1
====================================
```

I believe the above covers the re-factored code paths in `scheduleReduction`.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: Inline predicate today returns a True bool if there's no for loops, but probably should return the thread predicate now. I think that wasn't done before because of some other bug that Naoya fixed recently, CC @naoyam . I think it should be safe to change inline predicate. After we do that we should merge this else with your new else, then just predicate how we adding the expression with `for_loops_.empty()`:
```.cpp
if (for_loops_.empty()){
  loop_replacement_map_.insert({expr, inline_ite});
} else {
  for_loops_.back()->body().insert_before(expr, inline_ite);
  for_loops_.back()->body().erase(expr);
```

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/scheduler.cpp]: nit:
`has_outer_axis` or `has_iter_axis`

**Review by csarofeen — CHANGES_REQUESTED** [review]
Overall looks good to me, minor change requests.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: I don't remember why it wasn't before, shouldn't the thread predicate be also used for shared memory?

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: I don't know, I defer to your judgement on this one.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: I tried to see what would happen with shared memory, but I'm having different errors. First of all, we don't support allocations of 0-dim tensors (#508). I also hit another error around syncthreads insertion (#507).

For the mean time, I still think we should use thread predication both for shared and global memories, so please update the PR accordingly, @shmsong

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: Refactored this part and changed `getInlinePredicate` to return thread predicate instead of `true` when at top level. Did a quick check on generated kernel in the tests, didn't seem to produce side effects other than some equivalent re-ordering of generated predicates. Also reviewed places where `getInlinePredicate` gets used and this should be safe replacement at this point.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:61]: @naoyam Thanks for the suggestion. Added `MemoryType::Shared` in the predicate logic. #508 and #507 could be taken care of in a different PR and thanks for tracking these issues.

File: torch/csrc/jit/codegen/cuda/scheduler.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/scheduler.cpp]: Renamed as `has_iter_axis`. Thanks!

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp:57]: Just to make sure I'm reading this correctly. This is if there is a 0-Dim tensor, and the memory is global or shared, there may be multiple threads trying to write to this location. However, `getThreadPredicate` will determine who is actually supposed to write to that location.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:57]: That should be right.

Thinking about this again, wouldn't it make more sense to remove `!for_loops_.empty()`? I.e.,
```
const bool should_predicate = out_tv->memoryType() == MemoryType::Global ||
    out_tv->memoryType() == MemoryType::Shared;
```

Having thread predicates should be always safe. We can just omit them for local memory, whether expressions are inside loops or not, as it won't affect correctness.

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Please remove "except for top level expressions"

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp]: Updated Thanks!

File: torch/csrc/jit/codegen/cuda/lower_unroll.cpp
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:57]: This is great point, thanks! 

I actually wanted to understand more before committing a change. I understand that if output is local then we don't care about thread in terms of correctness but not sure if that's the whole story, mainly because I'm still brushing up on GPUs. 

Say for example, the op reads global memory, would this predicate have perf implications, or are we going to generate code using similar pattern?

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:57]: Yeah, that could have some perf overheads. Not sure how much difference we would see in terms of performance. I think either is fine for now.

Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/lower_unroll.cpp:57]: We'll postpone this improvement to the next few PRs that'd revisit this part. Thanks for the great suggestions! 👍

File: test/test_jit_cuda_fuser.py
Comment (jjsjann123) [inline test/test_jit_cuda_fuser.py:775]: Note to myself that we should update parser to enable sum to scalar for the overload of `torch.sum`.

**Pull request merged**
